ToolChain/Compiler/Compiler.py

```diff
@@ -0,0 +1,1783 @@
+    pointerList = []             # List containing variables that are pointers
+    charList = []                # List containing variables that are chars
+    quoteFlag = ""                # Keeps track of whether we used single or double quote to declare a char variable
+
+    def __init__(self, inputFile=None, outputFile=None):
+        """
+        This allows for simple initialisation of the Compiler. It will check if the input/output files are valid,
+        and then call the parseFile method to begin the compiling process.
+        :param inputFile: inputFile: str, name of file to read from
+        :param outputFile: str, name of file that will be created and written to
+        :return:
+        """
+
+        if type(inputFile) is not str or len(inputFile) is 0:
+            # File is invalid
+            raise ValueError("Assembler error - Invalid input file selected")
+        if type(outputFile) is not str or len(outputFile) is 0:
+            # File is invalid
+            raise ValueError("Assembler error - Invalid output file selected")
+
+        self.readFile(inputFile, outputFile)
+
+    def readFile(self, inputFile, outputFile):
+        """
+        Initializes parsing process. Opens the input file to read from, and initializes the output file we will create.
+        We then read each line and feed every character into the parse method individually. Once done, we close the
+        input/output files.
+        :param inputFile: str, name of file to read from
+        :param outputFile: str, name of file that will be created and written to
+        :return:
+        """
+
+        try:
+            file = open(inputFile, mode="r")
+            inputFile = file.readlines()
+        except OSError as e:
+            raise OSError("Couldn't open file {}".format(inputFile))
+        try:
+            output = open(outputFile, mode="w")
+        except OSError as e:
+            raise OSError("Couldn't open file {}".format(outputFile))
+
+        for line in inputFile:
+            self.lineno += 1
+            for x in line:
+                self.parse(x, output)
+
+        output.write("end:\n")
+
+        if self.currentMethod != "":
+            raise ValueError("Missing closing curly brace for end of method/if/while.")
+
+        try:
+            file.close()
+            output.close()
+        except OSError as e:
+            raise OSError("Couldn't close file.")
+
+    def parse(self, char, output):
+        """
+        Receives characters one by one and makes use of states to determine the next appropriate action. Each state has
+        its own restrictions based on what input it expects next. This parse method is essentially the controller that
+        directs input.
+        :param char: char, Individual character read from input file
+        :param output: file, output file to write to
+        :return:
+        """
+
+        if self.state == 0:
```

Comment:
The various states should be defined in constants with appropriate naming. This would make code easier to read.

Reply:
Will do